### PR TITLE
223-test: Update screenshots, improve tests

### LINKED ISSUE #223: Check design changes in main & update Playwright screenshots
Image

### LINKED ISSUE #232: Docs: Typical Playwright use cases for beginners
Please update the doc: https://github.com/rolling-scopes/site/blob/main/readme/end-to-end-testing.md

Example:
-----
# End-to-End Testing with Playwright

The project ensures that each new feature, once approved, is merged into the main branch and deployed to production. However, visual bugs (missing elements, shifted images) couldn't always be quickly detected manually. To address this, end-to-end testing with Playwright was implemented.

## How It Works

Before pushing changes to GitHub, developers can run end-to-end tests with `npx playwright test`. Playwright starts a server with the modified code, takes full-page screenshots, and compares them with reference images. Discrepancies result in test failures, with comparison results saved in the `src/__tests__/visualTesting/results` folder, helping identify and fix issues.

image

### Detecting Visual Bugs

Run `npx playwright test` to detect visual discrepancies between the current state and reference images. This is useful for catching issues like missing elements or layout shifts.

### Updating Reference Images

When intentional design changes cause test failures, delete outdated reference images (e.g., `src/__tests__/visualTesting/navbar.spec.ts-snapshots`) and rerun the tests with `npx playwright test` to create new references.

### Reviewing Test Results

Launch the Playwright UI with `npx playwright test --ui` to easily review test results and debug issues.

image

## Handling Intentional Changes

If intentional changes affect the site's appearance, update the reference images by deleting the old ones and rerunning the tests. The library will create new reference samples.

## User Interface for Testing

For a user-friendly way to run and check tests, use the Playwright UI with:
```sh
npx playwright test --ui
```

### Example

If you notice a discrepancy in the navbar's appearance, the test UI will highlight the differences and provide a visual comparison to help you identify and fix the issue.

## Dealing with Errors

When tests fail, review the error messages and comparison images in the `src/__tests__/visualTesting/results` folder to understand what went wrong. Common issues include layout shifts, missing elements, or unintended changes in styling.

### PR DESCRIPTION
**Pull request by ramzes-84**
## What type of PR is this? (select all that apply)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 🚧 Breaking Change
- [x] 🧑‍💻 Code Refactor
- [x] 📝 Documentation Update

## Description

- improve Playwright (PW) tests
- remove separate tests for header and footer 
- add Community page's test
- add Courses page's test
- update screenshots
- add mask areas for changeable content (pink blocks in the screenshots)
- replace prod server with dev server for testing
- update image formation in dev environment (remove lazy loading for tests)
- update PW version
- add custom css for PW better testing
- update end-to-end-testing readme

## Related Tickets & Documents

- Related Issue #223 
- Closes #223 
- Related Issue #232 
- Closes #232 
 
## Screenshots, Recordings

_Please replace this line with any relevant images for UI changes._

## Added/updated tests?

- [x] 👌 Yes
- [ ] 🙅‍♂️ No, because they aren't needed
- [ ] 🙋‍♂️ No, because I need help

### CODE REVIEW

File: readme/end-to-end-testing.md
Comment (Quiddlee) [inline readme/end-to-end-testing.md:1]: Well done! Now it's become way more better

File: src/__tests__/visualTesting/custom.css [hunk 1]
```diff
@@ -0,0 +1,3 @@
+nav {
+  position: static !important;
```
Comment (Quiddlee) [inline src/__tests__/visualTesting/custom.css:2]: Could you explain why is this necessary?

File: readme/end-to-end-testing.md [hunk 2]
```diff
@@ -1,25 +1,54 @@
-# Reasons for Implementation
+## 🚨Reasons for Implementation
```
Comment (dzmitry-varabei) [inline readme/end-to-end-testing.md:1]: Maybe change the title to something like “End-to-End Testing with Playwright”?

**Comment by dzmitry-varabei** [comment]
@ramzes-84 thanks a lot!

File: readme/end-to-end-testing.md (hunk 2 above)
Comment (ramzes-84) [reply readme/end-to-end-testing.md:1]: Like this?

# End-to-End Testing with Playwright

## 🚨Reasons for Implementation

Or replace `Reasons for Implementation` with `End-to-End Testing with Playwright`?

File: src/__tests__/visualTesting/custom.css (hunk 1 above)
Comment (ramzes-84) [reply src/__tests__/visualTesting/custom.css:2]: You have a sharp eye! Now it is not necessary since I've deleted separate test for footer section. I will delete custom css, thank you.
Just for clarification for others and sharing experience: sticky `nav` section covered footer on screenshots. Changing this css rule fixed this issue.

footer-Mobile-Chrome-win32

**Review by natanchik — COMMENTED** [review]
Don't we need something like this?
image

File: readme/end-to-end-testing.md [hunk 3]
```diff
@@ -1,25 +1,87 @@
...
+
+- to update one or several screenshots only, please delete outdated reference images from the according folder (`src/__tests__/visualTesting/[NAME_OF_THE_TEST].spec.ts-snapshots`) and rerun tests again. On testing without samples, Playwright firstly will create them from your code.
+
+## 🆕Handling Intentional Changes
+
+If intentional changes affect the site's appearance, update the reference images by deleting the old ones and rerunning the tests. The library will create new reference samples.
+
+## 🆘Troubleshooting and Errors
+
+### 🗐Screenshot Comparison Failed
```
Comment (natanchik) [inline readme/end-to-end-testing.md:81]:
```suggestion
### 🗐 Screenshot Comparison Failed
```

**Review by natanchik — APPROVED** [review]
Looks good 👍

**Review by Quiddlee — APPROVED** [review]

File: src/__tests__/visualTesting/community.spec.ts-snapshots/community-Mobile-Safari-win32.png
Comment (SpaNb4) [inline src/__tests__/visualTesting/community.spec.ts-snapshots/community-Mobile-Safari-win32.png:1]: Have you looked through the playwright issues on github to see if it's ok that we're seeing such “artifacts” for Safari?

**Comment by SpaNb4** [comment]
Do we have a ticket for integrating playwright into our ci/ci pipilene?

**Comment by ramzes-84** [comment]
> Do we have a ticket for integrating playwright into our ci/cd pipilene?

Not yet. And, in my opinion, it is too early to place testing on the server side. While I was doing this task on the weekend the design changed twice.
So let's normalize design matters and then configure CI/CD.

**Comment by ansivgit** [comment]
Please, fix PR name

File: src/__tests__/visualTesting/community.spec.ts-snapshots/community-Mobile-Safari-win32.png
Comment (ramzes-84) [reply src/__tests__/visualTesting/community.spec.ts-snapshots/community-Mobile-Safari-win32.png:1]: Unfortunately, I couldn't find the solution for this issue. The styles in the Webkit are the same as in Chrome, but in Webkit font is thinner.

image

image

File: readme/end-to-end-testing.md [hunk 4]
```diff
@@ -1,25 +1,103 @@
...
+
+```
+npx playwright test --update-snapshots
+```
+
+- to update one or several screenshots only, please delete outdated reference images from the according folder (`src/__tests__/visualTesting/[NAME_OF_THE_TEST].spec.ts-snapshots`) and rerun tests again. On testing without samples, Playwright firstly will create them from your code.
+
+### 👀 Real-time Browsing via Specified Engine
+
+You can even browse the websites using current engine. To do that, please use command:
```
Comment (SpaNb4) [inline readme/end-to-end-testing.md:77]: `You can browse websites using a specific rendering engine`

File: readme/end-to-end-testing.md [hunk 5]
```diff
@@ -1,25 +1,103 @@
...
+
+```
+npx playwright open -b webkit rs.school
```
Comment (SpaNb4) [inline readme/end-to-end-testing.md:80]: Wouldn't it be better to specify a local address here?

Comment (ramzes-84) [reply readme/end-to-end-testing.md:80]: It is just for example. I hope developers will replace server with one they need.

Comment (ramzes-84) [reply readme/end-to-end-testing.md:80]: Specifying local server is more confusing. We don't know wether the local server is running or is it using the same port. And then for those, who are "in the tank" we should add the instruction to run local service firstly and make sure than it has a `5173` port and so on. I think this command example is the best instruction.
But you can order me to do so as a code owner)

**Review by SpaNb4 — APPROVED** [review]
